records/README.md

```diff
@@ -0,0 +1,220 @@
+
+IPRS records are expressed as [merkledag](../merkledag) objects. This means that the records are linked authenticated data structures, and can be natively replicated over IPFS itself and other merkledag distribution systems.
+
+The objects:
+
+- A `Record` object expresses a value, a validity scheme, and validity data.
+- A `Signature` object could be used to sign and authenticate a record.
+- An `Encryption` object could be used to encrypt a record.
+
+```go
+Record Node {
+  Scheme   Link // link to a validity scheme
+  Value    Link // link to an object representing the value.
+  Validity Data // data needed to satisfy the validity scheme
+}
+```
+
+To achieve good performance, record storage and transfer should bundle all the necessary objects and transmit them together. While "the record object" is only one of the dag objects, "the full record" means a bundle of all objects needed to fully represent, verify, and use the record. (This recommendation does not necessarily include data that records _describe_, for example an ipfs provider record (which signals to consumers that certain data is available) would not include the data itself as part of "the full record").
+
+### The Interface
+
+The IPRS interface is below. It has a few types and functions. We use the Go language to express it, but this is language agnostic.
+
+```go
+// Record is the base type. user can define other types that
+// extend Record.
+type Record struct {
+  Scheme    Link // link to the validity scheme
+  Signature Link // link to a cryptographic signature over the rest of record
+  Value     Data // an opaque value
+}
+
+// Validator is a function that returns whether a record is valid.
+// Users provide their own Validator implementations.
+type Validator func(r *Record) (bool, error)
+
+// Marshal/Unmarshal specifies a way to code the record
+type Marshal(r *Record) ([]byte, error)
+type Unmarshal(r *Record, []byte) (error)
+```
+
+### Interface Example
+
+For example, Alice and Bob earlier could use the following interface:
+
+```protobuf
+type Record struct {
+  Scheme    Link // link to the validity scheme
+  Signature Link // link to a cryptographic signature over the record
+  Expires   Data // datetime at which record expires
+  Value     Data // an opaque value
+}
+
+
+func Validator(r *Record) (bool, error) {
+  authorKey := recordSigningKey(r)
+
+  // always check the signature first
+  sigok := authorKey.Verify(r.Signature, signablePart(r))
+  if !sigok {
+    return false, errors.New("invalid signature. forged record?")
+  }
+
+  // check the expiration.
+  if r.Expires < time.Now() {
+    return false, errors.New("record expired.")
+  }
+
+  return true, nil
+}
+
+func Marshal(r *Record) ([]byte, error) {
+  return recordProtobuf.Marshal(r)
+}
+
+func Unmarshal(r *Record, d []byte) (error) {
+  return recordProtobuf.Unmarshal(r, d)
+}
+```
+
+## Example Record Types
+
+For ease of use, IPRS implementations should include a set of common record types:
```

Comment:
If all of these records require a PKI, it seems sensible to at least suggest a common infrastructure, and point out that participating nodes are free to refuse to host or transmit records that don't have valid signatures in that infrastructure.  This might be a good place to spec out the keyspace limitations as well, since requiring IPRS keys to start with a multihash for the signing public key is how we handle namespacing keys to avoid attackers maliciously clobbering your valid records.  So having a “use PK-namespaced IPRS keys and records signed using PKI $X” limitation seems like a reasonable requirement to me, although it can be a separate layer on top of the unsigned base protocol if you like.